src/pages/studyView/addChartButton/geneLevelSelection/GeneLevelSelection.tsx

```diff
@@ -0,0 +1,199 @@
+import { GeneReplacement } from 'shared/components/query/QueryStore';
+import OQLTextArea, {
+    GeneBoxType,
+} from 'shared/components/GeneSelectionBox/OQLTextArea';
+import classnames from 'classnames';
+import MobxPromise from 'mobxpromise';
+import { MakeMobxView } from 'shared/components/MobxView';
+import LoadingIndicator from 'shared/components/loadingIndicator/LoadingIndicator';
+import ErrorMessage from 'shared/components/ErrorMessage';
+
+export interface IGeneLevelSelectionProps {
+    molecularProfileOptionsPromise: MobxPromise<
+        {
+            value: string[];
+            label: string;
+            profileName: string;
+            description: string;
+        }[]
+    >;
+    submitButtonText: string;
+    onSubmit: (charts: GenomicChart[]) => void;
+}
+
+@observer
+export default class GeneLevelSelection extends React.Component<
+    IGeneLevelSelectionProps,
+    {}
+> {
+    @observable private geneInput: string = '';
+    @observable private validGenes: Gene[] = [];
+    @observable private isQueryInvalid: boolean = true;
+    @observable private hasOQL: boolean = false;
+    @observable private selectedProfileOption?: {
+        value: string[];
+        label: string;
+        profileName: string;
+        description: string;
+    };
+
+    public static defaultProps = {
+        disableGrouping: false,
+    };
+
+    @autobind
+    @action
+    private onAddChart() {
+        if (this.selectedOption !== undefined) {
+            const charts = this.validGenes.map(gene => {
+                return {
+                    name:
+                        gene.hugoGeneSymbol +
+                        ': ' +
+                        this.selectedOption!.profileName,
+                    description: this.selectedOption!.description,
+                    patientAttribute: false,
+                    molecularProfileIds: this.selectedOption!.value,
+                    hugoGeneSymbol: gene.hugoGeneSymbol,
+                };
+            });
+            this.props.onSubmit(charts);
+        }
+    }
+
+    @autobind
+    @action
+    private handleSelect(option: any) {
+        if (option && option.value) {
+            this.selectedProfileOption = option;
+        }
+    }
+
+    @computed
+    private get selectedOption() {
+        if (this.selectedProfileOption !== undefined) {
+            return this.selectedProfileOption;
+        }
+        if (this.props.molecularProfileOptionsPromise.isComplete) {
+            return this.props.molecularProfileOptionsPromise.result![0];
+        }
+        return undefined;
+    }
+
+    @autobind
+    @action
+    private updateSelectedGenes(
+        oql: {
+            query: SingleGeneQuery[];
+            error?: { start: number; end: number; message: string };
+        },
+        genes: {
+            found: Gene[];
+            suggestions: GeneReplacement[];
+        },
+        queryStr: string
+    ) {
+        this.hasOQL = false;
+        this.isQueryInvalid =
```

Comment:
not sure but i would think something like isQueryInvalid should be a computed property derived from these things.

Reply:
it seems like there are multiple instance properties being set in this action which ought to be computations.